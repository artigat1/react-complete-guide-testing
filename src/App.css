html,
body {
    height: 100%;
    color: #61dafb;
    background: #333;
}

h1 {
    margin: 20px 0 15px 0;
}

#main {
    width: 600px;
    margin: 0 auto;
    text-align: left;
}

ul {
    list-style: none;
    margin: 0;
    padding: 0;
}

ul > li:hover {
    background-color: #efefef;
}

input {
    padding: 0.75em;
    width: 50%;
    margin: 0.75em 0;
}

form {
    margin-bottom: 1em;
}

button {
    margin: 0.5em;
    padding: 0.75em;
    background-color: teal;
    color: white;
    border: green;
}

button:hover {
    border: #61dafb;
    color: #61dafb;
    cursor: pointer;
}

.done {
    color: red;
    text-decoration: line-through;
}

.undone {
    color: #61dafb;
}

.icon {
    margin: 6px 10px 6px 0;
}

.error {
    color: red;
    font-size: 0.9em;
    margin: 0;
    padding: 0;
}
