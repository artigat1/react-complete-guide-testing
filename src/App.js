import React, { Component } from 'react';

import './App.css';
import TodoHeader from './components/Header/Header';
import TodoList from './components/TodoList/TodoList';
import TodoForm from './components/Form/Form';

class App extends Component {
    state = {
        todoItems: this.props.initItems
    };

    constructor(props) {
        super(props);

        this.addItem = this.addItem.bind(this);
        this.removeItem = this.removeItem.bind(this);
        this.markTodoDone = this.markTodoDone.bind(this);
    }

    addItem(todoItem) {
        const items = this.state.todoItems;
        items.unshift({
            index: this.state.todoItems.length + 1,
            value: todoItem.newItemValue,
            done: false
        });
        this.setState({ todoItems: items });
    }

    removeItem(itemIndex) {
        const items = this.state.todoItems;
        items.splice(itemIndex, 1);
        this.setState({ todoItems: items });
    }

    markTodoDone(itemIndex) {
        const items = this.state.todoItems;
        const todo = { ...items[itemIndex] };
        items.splice(itemIndex, 1);
        todo.done = !todo.done;
        todo.done ? items.push(todo) : items.unshift(todo);
        this.setState({ todoItems: items });
    }

    render() {
        return (
            <div id="main">
                <TodoHeader />
                <TodoForm addItem={this.addItem} />
                <TodoList
                    items={this.state.todoItems}
                    removeItem={this.removeItem}
                    markTodoDone={this.markTodoDone}
                />
            </div>
        );
    }
}

export default App;
