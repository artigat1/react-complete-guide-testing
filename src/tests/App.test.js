import React from 'react';
import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import App from '../App';

configure({ adapter: new Adapter() });

describe('<App />', () => {
    let wrapper;
    const todoItems = [];

    beforeEach(() => {
        todoItems.length = 0;
        todoItems.push({ index: 1, value: 'learn react', done: false });
        todoItems.push({ index: 2, value: 'Go shopping', done: true });
        todoItems.push({ index: 3, value: 'buy flowers', done: true });

        wrapper = shallow(<App initItems={todoItems} />);
    });

    it('should render correctly', () => {
        const component = renderer
            .create(<App initItems={todoItems} />)
            .toJSON();
        expect(component).toMatchSnapshot();
    });

    describe('markTodoDone', () => {
        it('should mark the correct todo item as done and move it to the last in the list', () => {
            const instance = wrapper.instance();
            const item = instance.state.todoItems.find(x => x.index === 1);
            item.done = false;

            expect(
                instance.state.todoItems.find(x => x.index === 1).done
            ).toEqual(false);

            instance.markTodoDone(0);

            expect(
                instance.state.todoItems.find(x => x.index === 1).done
            ).toEqual(true);
            expect(
                instance.state.todoItems.findIndex(x => x.index === 1)
            ).toEqual(todoItems.length - 1);
        });

        it('should mark the correct todo item as incomplete and move it to the top of the list', () => {
            const instance = wrapper.instance();
            const item = instance.state.todoItems.find(x => x.index === 1);
            item.done = true;

            expect(
                instance.state.todoItems.find(x => x.index === 1).done
            ).toEqual(true);

            instance.markTodoDone(0);

            expect(
                instance.state.todoItems.find(x => x.index === 1).done
            ).toEqual(false);
            expect(
                instance.state.todoItems.findIndex(x => x.index === 1)
            ).toEqual(0);
        });
    });

    describe('removeItem', () => {
        it('should remove the correct item from the list', () => {
            const instance = wrapper.instance();
            const item = instance.state.todoItems.find(x => x.index === 1);

            expect(item).toEqual(todoItems[0]);

            instance.removeItem(0);

            expect(
                instance.state.todoItems.find(x => x.index === 1)
            ).not.toBeDefined();
        });
    });

    describe('addItem', () => {
        it('should add a new item to the top of the list', () => {
            const instance = wrapper.instance();
            const newItem = { index: 4, value: 'new item', done: false };
            const expectedLength = todoItems.length + 1;

            expect(instance.state.todoItems.length).toEqual(todoItems.length);

            instance.addItem(newItem);

            expect(instance.state.todoItems.length).toEqual(expectedLength);
            expect(instance.state.todoItems[0].index).toEqual(newItem.index);
        });
    });
});
