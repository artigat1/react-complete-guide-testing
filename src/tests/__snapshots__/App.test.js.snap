// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render correctly 1`] = `
<div
  id="main"
>
  <h1>
    Todo list
  </h1>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="add a new todo..."
      type="text"
      value=""
    />
    <button
      className="btn btn-default"
      type="submit"
    >
      Add
    </button>
    
  </form>
  <ul
    className="list-group"
  >
     
    <li
      className="list-group-item "
    >
      <div
        className="undone"
      >
        <button
          className="close"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
         
        <span
          aria-hidden="true"
          className="glyphicon glyphicon-ok icon"
          onClick={[Function]}
        >
          learn react
        </span>
      </div>
    </li>
    <li
      className="list-group-item "
    >
      <div
        className="done"
      >
        <button
          className="close"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
         
        <span
          aria-hidden="true"
          className="glyphicon glyphicon-ok icon"
          onClick={[Function]}
        >
          Go shopping
        </span>
      </div>
    </li>
    <li
      className="list-group-item "
    >
      <div
        className="done"
      >
        <button
          className="close"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
         
        <span
          aria-hidden="true"
          className="glyphicon glyphicon-ok icon"
          onClick={[Function]}
        >
          buy flowers
        </span>
      </div>
    </li>
     
  </ul>
</div>
`;
